crates/client/db/src/rocksdb_options.rs

```diff
@@ -1,3 +1,50 @@
+//! # Rocksdb options
+//!
+//! We configure the rocksdb database in a very specific way. We want to guarantee a few things:
+//!
+//! ## Fault tolerance
+//!
+//! The madara node can be stopped at any time, and it should be able to recover when restarting.
+//! In particular, in case of a power/hardware failure, the database should not be corrupted beyond repair. We cannot
+//! run code that would allow us to gracefully shutdown in these cases.
+//!
+//! The way rocksdb ensures this is usually by using a WAL (write-ahead log) and doing a 2PC (two-phase commit) under
+//! the hood. There is however another way to achieve fault tolerance with rocksdb, which is by enabling the [atomic flush]
+//! option.
+//!
+//! ## Consistent view of the database
+//!
+//! The usual way it is achieved is by batching all of the changes into a single batch or a single rocksdb transaction so
+//! that they are applied to the database atomically.
+//!
+//! ## Multithreaded commit to the database
+//!
+//! This is main reason we differ from the default rocksdb configuration: rocksdb Transaction objects are not thread safe,
+//! and committing a transaction also isn't multithreaded at all. This means that if we want to write a whole block into the
+//! database, we would have to put all of the changes into a single WriteBatch and apply it all at once in a single-threaded
+//! fashion. Rocksdb has [WriteUnprepared transactions] which would help us avoid buffering all the changes before applying them,
+//! but the transaction object still cannot be passed to other threads, so that wouldn't work either.
+//!
+//! The way we work around that is by:
+//! - enabling atomic flushing
+//! - disabling the WAL, as we don't need it with atomic flushing
+//! - disabling auto flushing, so that a block cannot be written into the db half way
+//! - do not use any transactions, do all the writes directly to the database.
+//! - use explicit [Snapshot]s for reads: all reads use the latest snapshot; and between every block we do a new snapshot and drop the old one.
+//!   This ensures we don't read the database when a block has been half-way written
+//!
+//! That last point isn't yet implemented, as we need some changes to the rust rocksdb bindings, see [rust-rocksdb#937]. No issue has yet been
+//! found with this as our db is almost only append-only - but this should nonetheless be fixed. (FIXME)
+//!
+//! This configuration makes a lot of sense because we making a blockchain node. All of our db writes are very big, all or nothing and infrequent.
```

Comment:
This configuration makes a lot of sense in the context of a blockchain node.

Reply:
thank you very much